docs/src/developer/topics.md

```diff
@@ -0,0 +1,387 @@
+that inplace modification of Julia array elements is not safe as it would
+modify the original object that was assigned to that location, whereas in Flint
+inplace modification is highly desirable for performance reasons and is
+completely safe due to the fact that a copy was made when the assignment
+happened.
+
+We have developed over a period of many years a set of rules that maximise the
+performance benefit we get from our unsafe operators, whilst keeping the
+burden imposed on the programmer to a minimum. It has been a *very* difficult
+task to arrive at the set of rules we have whilst respecting correctness of our
+code, and it would be extremely hard to change any of them.
+
+### Arithmetic operations return a new object
+
+In order to make it easy for the Nemo developer to create a completely new
+object when one is needed, e.g. for accumulating values using unsafe operators,
+we developed the following rules.
+
+Whenever an arithmetic operation is used, i.e. `+`, `-`, `*`, unary minus and
+`^`, Nemo always returns a new object, in line with Julia. Naturally,
+`deepcopy` also makes a copy of an object which can be used in unsafe
+functions.
+
+Note that if `R` is a type and an element `a` of that type is passed to it,
+e.g. `R(a)` then, the Julia convention is that `a` will be returned rather
+than a copy of `a`. This convention ensures there is not an additional cost
+when coercing values that are already of the right type.
+
+We extend this convention to parent objects `R` and elements `a` of that
+parent. In particular, `R(a)` cannot be used to make a copy of `a` for use
+in an unsafe function if `R` is the parent of `a`.
+
+All other functions *may* also return the input object if they wish. In other
+words, the return value of all other functions is not suitable for use in
+an unsafe function. Only return values of arithmetic operations and `deepcopy`
+will be suitable for such use.
+
+This convention has been chosen to maximise performance of Nemo. Low level
+operations (where performance matters) make a new object, even if the result
+is the same arithmetically as one of the inputs. But higher level functions
+will not necessarily make a new object, meaning that they cannot be used with
+unsafe functions.
+
+### Aliasing rules
+
+We now summarise the aliasing rules used by Nemo and AbstractAlgebra. We are
+relatively confident by now that following these rules will result in correct
+code given the constraints mentioned above.
+
+* matrices are viewed as containers which may contain elements that alias one
+another. Other objects, e.g. polynomials, series, etc., are constructed from
+objects that do not alias one another, *even in part*.
+
+* standard unsafe operators, addeq!, mul!, addmul!, zero!, add!, are allowed
+to mutate their outputs iff the output is entirely under the control of the
+caller, i.e. it was created for the purpose of accumulation, but otherwise
+must not modify any inputs or outputs
```

Comment:
This sounds confusing, as `addeq!` etal. don't know about the context where they are used, their job is to mutate when possible. I believe you mean something like the following?
```suggestion
* standard unsafe operators, addeq!, mul!, addmul!, zero!, add!, are allowed to be used 
to mutate their outputs iff the output is entirely under the control of the
caller, i.e. it was created for the purpose of accumulation, but otherwise
must not be used.
```

Reply:
Yeah hopefully that is what is meant. I changed it.